### PR TITLE
[do not merge] Remove hardcoded AWS image urls and download images into assets

### LINKED ISSUE #5810: Remove HardCoded AWS Image URLs and Download Images into /assets
Throughout our codebase, we have a lot of hardcoded AWS image urls. [Here is an example](https://github.com/thepracticaldev/dev.to/pull/5799/files) of a few of them. There are even more examples sprinkled all over the views. 

Rather than updating the AWS bucket name everywhere which will break for many people's local integrations we should download these images and add them to the `/assets/images` directory so they can be served directly from the app. In addition, we should give those that we download pertinent names relating to what they are so when another community would like to spin up this codebase for their own purposes they can replace the images with ones that represent their use case.

### PR DESCRIPTION
**Pull request by seb1441**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
I removed hardcoded AWS image urls and downloaded them in the assets folder with appropriate names.
As a bonus, all the broken image urls/displaying in local that I noticed seem to be fixed.

### Notes:
- For cloudinary image urls, I used `cl_image` helper.
**IMPORTANT**: However, we will need to run `rake cloudinary:sync_static` to upload local images to cloudinary and push the generated files `.cloudinary.static` and `.cloudinary.static.trash`, which keeps track of what's been uploaded, in git. I did not include my files because I believe it's linked to my cloudinary account so it needs to be done with DEV's cloudinary account.
- There was one image, `octopus.png` which was too big for cloudinary (12.3 MB) so I had to reduce its size. https://thepracticaldev.s3.amazonaws.com/i/jl3vwgqxcx37vr45ln2o.png
- There are still AWS image urls left in `api_v0.yml`, `offline.html` and `README.md` but I assume that we need to leave them like that.

## Related Tickets & Documents
Closes #5810 

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

### CODE REVIEW

**Comment by seb1441** [comment]
@rhymes In my PR, I simply used `cl_image_tag` when the image url was using cloudinary and `image_tag` if not. But the criteria could be to use `cl_image_tag` when we want to do any image manipulation/transformation with cloudinary, otherwise `image_tag` as it's probably a bit more performant.

**Comment by rhymes** [comment]
Hi @thepracticaldev/sre - I tagged you for the review because this PR requires uploading static images to our Cloudinary account, it's not just about the code itself

**Comment by nickytonline** [comment]
@rhymes, do we have a script for to upload to Cloudinary? I looked in the issue and there does not appear to be one linked in the issue.

**Comment by rhymes** [comment]
@nickytonline it's builtin, it is mentioned in the initial description by @seb1441:

> IMPORTANT: However, we will need to run rake cloudinary:sync_static to upload local images to cloudinary and push the generated files .cloudinary.static and .cloudinary.static.trash, which keeps track of what's been uploaded, in git. I did not include my files because I believe it's linked to my cloudinary account so it needs to be done with DEV's cloudinary account.

File: app/views/pages/about.html.erb [hunk 1]
```diff
@@ -63,7 +63,7 @@
     <div class="about-rows">
       <div class="about-row">
         <div class="about-key">
-          <img src="<%= cloudinary "https://thepracticaldev.s3.amazonaws.com/i/b22fx2whfnyfmm05p856.jpeg" %>" width="250" alt="Ben Halpern" />
+          <%= image_tag "ben-halpern.jpeg", width: 250, alt: "Ben Halpern" %>
```
Comment (nickytonline) [inline app/views/pages/about.html.erb]: Shouldn't this be a `cl_image_tag` as well?

**Comment by nickytonline** [comment]
@seb1441, looks good for the `image_tag` conversions. I'm testing `cl_image_tag`. but just sorting out some issues when running `rake cloudinary:sync_static`. I'm getting `CloudinaryException: Invalid Signature xxx. String to sign - xxxx`. I've never used Cloudinary, so just reading through [their docs](https://cloudinary.com/documentation/upload_images#generating_authentication_signatures).

**Comment by nickytonline** [comment]
@seb1441, I also noticed that the about page is different in your branch. The latest in production only has a group picture of the founders and then a team photo below it. You'll need to update your branch with the latest from the main repository's master.

File: app/views/pages/about.html.erb (hunk 1 above)
Comment (nickytonline) [reply app/views/pages/about.html.erb]: Ignore this because as mentioned there is only a group photo of the founders now.

**Comment by seb1441** [comment]
@nickytonline I updated my branch with the latest from the main repository's master but my about page is still different from https://dev.to/about. Do you know why?

I also checked directly on github and it has the same about page as my branch: https://github.com/thepracticaldev/dev.to/blob/master/app/views/pages/about.html.erb

I'm a bit confused.

**Comment by rhymes** [comment]
hi @seb1441, the about page in production is loaded from the DB, though in dev defaults to the default one you can see

**Comment by nickytonline** [comment]
Any reason we can’t load the same about page from the DB for local development @rhymes? Also, if we keep the local dev about page, I imagine we don’t need to use `cl_image_tag` for those images.

**Comment by rhymes** [comment]
@nickytonline because it's no present in the seeded development DB. We could actually add it to the seed file BUT cloudinary images don't really work locally so it might not be worth it.

I think the about template is there only for `development` and new installations.

File: app/views/pages/shecoded.html.erb [hunk 2]
```diff
@@ -0,0 +1,289 @@
+<!DOCTYPE html>
```
Comment (maestromac) [inline app/views/pages/shecoded.html.erb]: This file was previously removed. I assume it came back because of the rebase.

Comment (seb1441) [reply app/views/pages/shecoded.html.erb]: Oops I thought I had removed it, done.

**Comment by rhymes** [comment]
Hi @seb1441, the file `app/views/podcast_episodes/_subscribe_buttons.html.erb` disappeared from master recently so you might have to rebase again and remove it from the conflict, sorry!

**Comment by seb1441** [comment]
Hi @rhymes, no problem. Done!

**Comment by nickytonline** [comment]
> @seb1441, I couldn't find it in the documentation, but when cl_image_tag is used in development mode, it seems to generate URL like image_tag does. Is this the expected behaviour? If so, the PR looks good. As mentioned, just update the branch with the latest from the main repository's master and this should be good to go.

The above is the only question I have left. From what I can tell, on my local, t seems to just act like `image_tag`. Is that the expected behaviour, i.e. no Cloudinary URL on my local?

**Comment by seb1441** [comment]
@nickytonline Sorry I skipped your question before.

The reason `cl_image_tag` acts as a normal `image_tag` (as in, the generated url does not point to cloudinary) is because I set `static_file_support` to false in development and test environments. That way, anyone will be able to see images even if they did not set their cloudinary credentials.

But on prod it will work. Example, this:
`cl_image_tag("jess-lee.jpg", width: 250, crop: :scale, fetch_format: "auto", flags: ["progressive"], quality: "auto", type: :asset, alt: "Jess Lee")`
will output
`Jess Lee`

If you wanna try it, you can set `static_file_support` to true for development in `config/cloudinary.yml`.

File: app/views/pages/about.html.erb [hunk 3]
```diff
@@ -81,7 +81,7 @@
       </div>
       <div class="about-row">
         <div class="about-key">
-          <img src="<%= cloudinary "https://thepracticaldev.s3.amazonaws.com/i/pm1d5npekd2j6ts2hcte.jpeg" %>" width="250" alt="Peter Frank" />
+          <%= cl_image_tag("peter-frank.jpg", width: 250, crop: :scale, fetch_format: "auto", flags: ["progressive"], quality: "auto", type: :asset, alt: "Peter Frank") %>
```
Comment (nickytonline) [inline app/views/pages/about.html.erb]: As @rhymes mentioned, the about page in prod is different, so we actually don' need this to be a `cl_image_tag`. Same for Peter's photo below.

Comment (seb1441) [reply app/views/pages/about.html.erb]: @nickytonline Changed to normal `image_tag`.

**Comment by rhymes** [comment]
Thanks for this, we need the SRE team to sign off on this but code wise it's fine! 

Great job @seb1441!

File: config/cloudinary.yml [hunk 4]
```diff
@@ -0,0 +1,11 @@
+development:
+  enhance_image_tag: true
+  static_file_support: false
+
+test:
+  enhance_image_tag: true
+  static_file_support: false
+
+production:
```
Comment (benhalpern) [inline config/cloudinary.yml]: At first glance, this is what could possibly be inconsistent with what we currently do? I'm not sure, just an inkling.

I can dive deeper later.

**Comment by citizen428** [comment]
Just commenting on here to re-surface this PR. What do we still need, what are the next steps?

**Comment by rhymes** [comment]
@citizen428 I think we need @thepracticaldev/sre team's to do a review of this, making sure there's nothing concerning and we also need their help in deploying it.

File: config/cloudinary.yml (hunk 4 above)
Comment (mstruve) [reply config/cloudinary.yml]: @benhalpern can you explain further what you mean here since it seems we want to get this pushed out

**Comment by rhymes** [comment]
Hi @seb1441, sorry for the late response. 

We've discussed this PR internally and we'd like to revive it and take it to completion, also in lieu to our generalization efforts as you well know. 

The main blocker (aside from conflicts due to our delayed response time) is that we don't plan to upload images to cloudinary. The flow for our website should always be:

1. add the images to the repo
2. tell cloudinary to fetch them and optimize them
3. show them to the user

Thus, we can't "sync" them to Cloudinary ahead of time.

If it's alright with you we'd like to "adopt" the PR and bring it to completion. Let us know what you think.

(cc. @Ridhwana which will likely be the team member working on finalizing this)

**Comment by seb1441** [comment]
Hi @rhymes, I understand. No problem, you can "adopt" my PR.

**Comment by mstruve** [comment]
Since this is going to be brought in house for completion I am going to close this until it can be updated and is ready to merge. 

Thanks @seb1441 for all your hard work on this!!!!

**Pull request closed without merging**